src/supervisor.rs

```diff
@@ -0,0 +1,614 @@
+        ordered_children_on_start(sup_order, children_specs0);
+
+    for failed_child_spec in pending_children {
+        start_result = start_result
+            .start_child(
+                ctx,
+                ev_notifier.clone(),
+                sup_runtime_name,
+                failed_child_spec,
+            )
+            .await;
+    }
+
+    match start_result {
+        StartResult::SuccessfulStart {
+            started_workers, ..
+        } => Ok(started_workers),
+        StartResult::FailedStart {
+            failed_worker_spec,
+            start_error,
+            prev_started_workers,
+            pending_workers,
+        } => {
+            let termination_result = terminate_prev_siblings(
+                &mut ev_notifier,
+                failed_worker_spec,
+                start_error,
+                pending_workers,
+                prev_started_workers,
+            )
+            .await;
+            Err(termination_result)
+        }
+    }
+}
+
+// terminate_children terminates each worker::Worker record. In case any of the
+// children fail to terminate a supervisor::TerminationError is returned.
+//
+async fn terminate_children<'a>(
+    ev_notifier: &mut EventNotifier,
+    runtime_children: impl Iterator<Item = Worker>,
+) -> Result<Vec<worker::Spec>, (Vec<worker::Spec>, TerminationError)> {
+    let mut children_specs = Vec::new();
+    let mut termination_errors: HashMap<String, Arc<worker::TerminationError>> = HashMap::new();
+    for runtime_child in runtime_children {
+        let child_runtime_name = runtime_child.runtime_name.to_owned();
+        let result = runtime_child.terminate().await;
+        match result {
+            Err((failed_child_spec, termination_err0)) => {
+                let termination_err = Arc::new(termination_err0);
+                ev_notifier
+                    .worker_termination_failed(child_runtime_name.clone(), termination_err.clone())
+                    .await;
+                termination_errors.insert(child_runtime_name, termination_err);
+                children_specs.push(failed_child_spec);
+            }
+            Ok(failed_child_spec) => {
+                ev_notifier
+                    .worker_terminated(child_runtime_name.clone())
+                    .await;
+                children_specs.push(failed_child_spec);
+            }
+        }
+    }
+    if termination_errors.is_empty() {
+        Ok(children_specs)
+    } else {
+        Err((children_specs, TerminationError(termination_errors)))
+    }
+}
+
+impl Spec {
+    pub fn new(
+        name: impl Into<String>,
+        children: Vec<worker::Spec>,
+        ev_notifier: EventNotifier,
+    ) -> Spec {
+        Spec {
+            name: name.into(),
+            ev_notifier,
+            children,
+
+            order: Order::LeftToRight,
+            strategy: Strategy::OneForOne,
+        }
+    }
+
+    async fn start_supervisor<'a>(
```

Comment:
This function should probably be `run_supervisor`, and should not return a `Supervisor` but rather just block until termination. Later, we can implement a `start_supervisor` that calls `run_supervisor` in a `tokio::task`, returning a `Supervisor` record that has an associated `join_handle` (the same way we do with the `Worker` API)